controllers/cloudstackfailuredomain_controller.go

```diff
@@ -130,25 +152,229 @@ func (r *CloudStackFailureDomainReconciliationRunner) ClearMachines() (ctrl.Resu
 	if err := r.K8sClient.List(r.RequestCtx, machines, client.MatchingLabels{infrav1.FailureDomainLabelName: r.ReconciliationSubject.Name}); err != nil {
 		return ctrl.Result{}, err
 	}
-	// Deleted CAPI machines for CloudStack machines found.
-	for _, machine := range machines.Items {
+	r.Log.Info(fmt.Sprintf("failuredomain delete: %d VM found in failuredomain.", len(machines.Items)))
+	if len(machines.Items) > 0 {
+		if result, err := checkClusterReady(r); err != nil {
+			return result, err
+		}
+	}
+
+	result, err := triggerEtcdClusterRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	result, err = triggerControlPlaneRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	result, err = triggerMachineDeploymentRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	if len(machines.Items) > 0 {
+		return r.RequeueWithMessage("failuredomain delete: FailureDomain still has machine(s) in it.", "failuredomain", r.ReconciliationSubject.Spec.Name)
+	}
+	return ctrl.Result{}, nil
+}
+
+func triggerEtcdClusterRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	etcdMachineFound := false
+	for _, machine := range machines {
 		for _, ref := range machine.OwnerReferences {
-			if ref.Kind == "Machine" {
-				machine := &clusterv1.Machine{}
-				machine.Name = ref.Name
-				machine.Namespace = r.ReconciliationSubject.Namespace
-				if err := r.K8sClient.Delete(r.RequestCtx, machine); err != nil {
+			if ref.Kind == kindEtcdadmCluster {
+				etcdMachineFound = true
+				// get etcdadmcluster by using machine's etcdadmcluster ownerReference
+				etcdadmCluster := &unstructured.Unstructured{}
+				etcdadmCluster.SetGroupVersionKind(schema.FromAPIVersionAndKind(ref.APIVersion, ref.Kind))
+				if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: machine.Namespace, Name: ref.Name}, etcdadmCluster); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				csMachineTemplateName, err := getCsMachineTemplateName(machine, etcdadmCluster, ref.Name)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+				newCsMachineTemplateName, err := cloneCsMachineTemplateWithNewName(r, machine.Namespace, csMachineTemplateName, ref.Name)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// patch etcdadmcluster to refer newly created cloudstack machine template, which will trigger an etcdadmcluster rolling upgrade.
+				if _, err := controllerutil.CreateOrPatch(r.RequestCtx, r.K8sClient, etcdadmCluster,
```

Comment:
What deletes the original machine template?

Reply:
cloudstack machine templates keeps, nobody remove the old ones. I think this is true for cluster upgrade.

Reply:
orignal machine template will not be deleted -- I think this is true for normal upgrade.

Reply:
Honestly, I think the problem this PR is aiming to solve needs to be discussed with all of the projects involved.

CAPI designers in particular may have a notion about failureDomain deletion, and if not, then we need to push them on why?

Either they fully support this feature or they don't. I don't like that we're in between two worlds.

We'll end up building something that half uses CAPI's failureDomain implementation and half doesn't.

Later CAPI may then fully commit or get out of this feature and where will that leave us?

Reply:
But also CAPI, EKS-A, and ETCD owners need to probably find a way to trigger a rolling update in a generic way.

I think any controller that manages CAPI Machines should standardize on the way they manage those machines.

Reply:
It kinds of like replicasets under deployment, each deployment changes will generate a new replicasets and it can be used for history record and roll back.

It does leak resource in etcd. The solution used in deployment is having a `revisionHistoryLimit` in deployment spec. 

So a similar solution for this is to have a `templateHistoryLimit` under somewhere (etcdadmcluster, capc's cloudstackmachinetemplate controller?)

SImply deleting the old cloudstackmachine template in above PR code is risky since cloudstackmachines are still referring them in `cloned-from`  labels.